File: es-examples/src/main/scala/org/eligosource/eventsourced/guide/StackableTraits.scala
/*
 * Copyright 2012-2013 Eligotech BV.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.eligosource.eventsourced.guide

import java.io.File

import akka.actor._

import org.eligosource.eventsourced.core._
import org.eligosource.eventsourced.journal.leveldb.LeveldbJournalProps

object StackableTraits extends App {
  implicit val system = ActorSystem("guide")

  // create a journal
  val journal: ActorRef = LeveldbJournalProps(new File("target/guide-2"), native = false).createJournal

  // create an event-sourcing extension
  val extension = EventsourcingExtension(system, journal)

  // event-sourced processor
  class Processor extends Actor { this: Emitter =>
    var counter = 0

    def receive = {
      case event => {
        // update internal state
        counter = counter + 1
        // print received event and number of processed events so far
        println("[processor] event = %s (%d)" format (event, counter))
        // send new event to destination channel
        emitter("destination") sendEvent ("processed %d events so far" format counter)
      }
    }
  }

  // channel destination
  class Destination extends Actor {
    def receive = {
      case event => {
        // print event received from processor via channel
        println("[destination] event = '%s'" format event)
      }
    }
  }

  // create and register event-sourced processor
  val processor: ActorRef = extension.processorOf(Props(new Processor with Emitter with Eventsourced { val id = 1 } ))

  // create channel destination
  val destination: ActorRef = system.actorOf(Props(new Destination with Receiver with Confirm))

  // create and register a named channel
  extension.channelOf(DefaultChannelProps(1, destination).withName("destination"))

  // recover registered processors by replaying journaled events
  extension.recover()

  // send event message to processor (will be journaled)
  processor ! Message("foo")

  // wait for all messages to arrive (graceful shutdown coming soon)
  Thread.sleep(1000)

  // then shutdown
  system.shutdown()
}

File: es-examples/src/main/java/org/eligosource/eventsourced/guide/japi/StackableTraits.java
/*
 * Copyright 2012-2013 Eligotech BV.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.eligosource.eventsourced.guide.japi;

import java.io.File;

import akka.actor.*;

import org.eligosource.eventsourced.core.*;
import org.eligosource.eventsourced.journal.leveldb.LeveldbJournalProps;

public class StackableTraits {
    public static class Processor extends UntypedEventsourcedEmitter {
        private int counter = 0;

        @Override
        public int id() {
            return 1;
        }

        @Override
        public void onReceive(Object event) throws Exception {
            counter = counter + 1;
            System.out.println(String.format("[processor] event = %s (%d)", event, counter));
            emitter(1).sendEvent(String.format("processed %d event messages so far", counter), getSelf());
        }
    }

    public static class Destination extends UntypedConfirmingReceiver {
        @Override
        public void onReceive(Object event) throws Exception {
            System.out.println(String.format("[destination] event = %s", event));
        }
    }

    public static void main(String... args) throws Exception {
        final ActorSystem system = ActorSystem.create("guide");

        final ActorRef journal = LeveldbJournalProps.create(new File("target/guide-2-java")).withNative(false).createJournal(system);
        final EventsourcingExtension extension = EventsourcingExtension.create(system, journal);

        final ActorRef destination = system.actorOf(Props.create(Destination.class));
        final ActorRef channel = extension.channelOf(DefaultChannelProps.create(1, destination), system);
        final ActorRef processor = extension.processorOf(Props.create(Processor.class), system);

        extension.recover();

        processor.tell(Message.create("foo"), null);

        Thread.sleep(1000);
        system.shutdown();
    }
}
